### DISCUSSION TITLE
How to send multiple search parameters?

### DISCUSSION THREAD

**Discussion by Laanan** [mjs/imapclient#433] [Q&A] [answered]
Hi folks, 

First, thanks, Menno, for opening this discussion thread!

I am making a CUI based IMAP client with py_cui (https://github.com/Laanan/Slither/tree/master), and I am struggling to understand how to list messages with multiple flags. 

For example, let's say I want to display the 50 newest messages, both seen and unseen. I have tried:

` def get_messages(self, folder):
 
          self.client.select_folder(folder)
          messages = self.client.search([ u'SEEN', u'UNSEEN', 1:50])
          message_list = []
          for uid, message_data in self.client.fetch(messages, "RFC822").items():
              email_message = email.message_from_bytes(message_data[b"RFC822"])
              message_list.append(email_message.get("Subject"))
 
          return message_list;
`
This doesn't work...

I am also hoping someone can direct me to further reading/a good book for helping me understand the methods/best practices/strategies for making an IMAP client function the way a user would expect. Something a little more in depth than a one off tutorial, and a little more tutorialy than the, somewhat staid, RFC document. 

Any advice would be greatly appreciated -- thank you!

-- Matt

**Comment by shoaib30** [comment]
You would need to use the **sort** command instead of search and slice the list to first 50. The sort command takes a sort criteria and a set of flags similar to search. But sort does not work in all IMAP servers including GMail as mentioned in [Stackoverflow](https://stackoverflow.com/a/5641178)

An alternative (as suggested in the above link) would be to search for ALL and fetch the BODY[HEADER] or ENVELOPE only, with which you can use the date to sort the mails in python. Once sorted you may slice the list and fetch the entire mail for the first 50. This method would certainly be much slowly with large mail boxes

**Comment by shoaib30** [comment, marked as answer]
Sorry I missed the main part of the question

To pass multiple search criteria with an OR condition you can do the following
`connection.search(['OR', 'UNSEEN', 'SEEN'])`

this link helped me get familiar with the search options :
[IMAP Search Criteria gist](https://gist.github.com/martinrusev/6121028)

or if you prefer you can refer to the [RFC ](https://tools.ietf.org/html/rfc3501#section-6.4.4)Document

**Reply by Laanan** [reply]
Hello,

Thanks for all of the super helpful information! 

One follow up question: is it advisable to use the standard library 'email' package alongside the 'imapclient' package? Or is it redundant?

**Reply by shoaib30** [reply]
Yes it is as per the documentation to use the standard email package as follows 

[IMAPClient Docs](https://imapclient.readthedocs.io/en/2.2.0/concepts.html#working-with-fetched-messages)
